Modix.Data/Repositories/PromotionCampaignRepository.cs

```diff
@@ -0,0 +1,176 @@
+            : base(modixContext, promotionActionEventHandlers) { }
+
+        /// <inheritdoc />
+        public Task<IRepositoryTransaction> BeginCreateTransactionAsync()
+            => _createTransactionFactory.BeginTransactionAsync(ModixContext.Database);
+
+        /// <inheritdoc />
+        public Task<IRepositoryTransaction> BeginCloseTransactionAsync()
+            => _closeTransactionFactory.BeginTransactionAsync(ModixContext.Database);
+
+        /// <inheritdoc />
+        public async Task<long> CreateAsync(PromotionCampaignCreationData data)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+
+            var entity = data.ToEntity();
+
+            await ModixContext.PromotionCampaigns.AddAsync(entity);
+            await ModixContext.SaveChangesAsync();
+
+            entity.CreateAction.CampaignId = entity.Id;
+            await ModixContext.SaveChangesAsync();
+
+            await RaisePromotionActionCreatedAsync(entity.CreateAction);
+
+            return entity.Id;
+        }
+
+        /// <inheritdoc />
+        public Task<bool> AnyAsync(PromotionCampaignSearchCriteria searchCriteria)
+            => ModixContext.PromotionCampaigns.AsNoTracking()
+                .FilterBy(searchCriteria)
+                .AnyAsync();
+
+        /// <inheritdoc />
+        public async Task<IReadOnlyCollection<PromotionCampaignSummary>> SearchSummariesAsync(PromotionCampaignSearchCriteria searchCriteria)
+            => await ModixContext.PromotionCampaigns.AsNoTracking()
+                .FilterBy(searchCriteria)
+                .Select(PromotionCampaignSummary.FromEntityProjection)
+                .ToArrayAsync();
+
+        /// <inheritdoc />
+        public Task<PromotionCampaignDetails> ReadDetailsAsync(long campaignId)
+            => ModixContext.PromotionCampaigns.AsNoTracking()
+                .Where(x => x.Id == campaignId)
+                .Select(PromotionCampaignDetails.FromEntityProjection)
+                .FirstOrDefaultAsync();
+
+        /// <inheritdoc />
+        public async Task<bool> TryCloseAsync(long campaignId, ulong closedById, PromotionCampaignOutcome outcome)
+        {
+            var entity = await ModixContext.PromotionCampaigns
+                .Where(x => x.Id == campaignId)
+                .FirstOrDefaultAsync();
+
+            if ((entity == null) || (entity.CloseActionId != null))
+                return false;
+
+            entity.Outcome = outcome;
+            entity.CloseAction = new PromotionActionEntity()
+            {
+                GuildId = entity.GuildId,
+                Type = PromotionActionType.CampaignClosed,
+                Created = DateTimeOffset.Now,
+                CreatedById = (long)closedById,
+                CampaignId = entity.Id
+            };
+            await ModixContext.SaveChangesAsync();
+
+            await RaisePromotionActionCreatedAsync(entity.CreateAction);
```

Comment:
This should be "entity.CloseAction", no?

Reply:
Ayup.

Reply:
Ayup